fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerPlaceBlockEvents.java

```diff
@@ -0,0 +1,72 @@
+package net.fabricmc.fabric.api.event.player;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ItemPlacementContext;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public class PlayerPlaceBlockEvents {
+	public static final Event<Before> BEFORE = EventFactory.createArrayBacked(Before.class,
+			(listeners) -> ((context) -> {
+				for (Before listener : listeners) {
+					ActionResult result = listener.beforeBlockPlace(context);
+					if(result != ActionResult.PASS) {
+						return result;
+					}
+				}
+				return ActionResult.PASS;
+			})
+	);
+
+	public static final Event<After> AFTER = EventFactory.createArrayBacked(After.class,
+			(listeners) -> ((world, entity, pos, state) -> {
+				for (After listener : listeners) {
+					listener.afterBlockPlace(world, entity, pos, state);
+				}
+			})
+	);
+
+	public static final Event<Cancelled> CANCELLED = EventFactory.createArrayBacked(Cancelled.class,
+			(listeners) -> ((world, entity, pos, state) -> {
+				for (Cancelled listener : listeners) {
+					listener.cancelledBlockPlace(world, entity, pos, state);
+				}
+			})
+	);
+
+	public static final Event<Allow> ALLOW = EventFactory.createArrayBacked(Allow.class,
+			(listeners) -> ((entity) -> {
+				for (Allow listener : listeners) {
+					if(!listener.allowBlockPlace(entity)) {
+						return false;
+					}
+				}
+				return true;
+			})
+	);
+
+	@FunctionalInterface
+	public interface Before {
+		ActionResult beforeBlockPlace(ItemPlacementContext context);
```

Comment:
I wonder if this is the optimal parameter choice... An AE2 cable would for example place the AE2 multipart block, which is really not obvious from the `ItemPlacementContext`.